### ISSUE TITLE
[JWT] TokenValidationParameters RoleClaimType Not Fully Respected

### ISSUE THREAD

**Issue by clauzf** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#1214]
Original issue: https://github.com/aspnet/AspNetCore/issues/11789

### Describe the bug
After passing value to RoleClaimType. It can no longer pass `Authorize(Roles="")`.

### To Reproduce
1. Using this version of ASP.NET Core '2.2'
2. Run this code

Startup.cs
```
// Inject token service
services.AddScoped<ITokenService, TokenService>();

// Inject appsettings.json
var tokenConfigs = Configuration.GetSection("tokenManagement");
services.Configure<TokenConfigs>(tokenConfigs);

// JWT authorization setup
var token = tokenConfigs.Get<TokenConfigs>();
services.AddAuthentication(x =>
{
    x.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
    x.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
}).AddJwtBearer(x =>
{
    x.RequireHttpsMetadata = false;
    x.SaveToken = true;
    x.TokenValidationParameters = new TokenValidationParameters
    {
        ValidateLifetime = true,
        ValidateIssuer = true,
        ValidateAudience = true,
        ValidateIssuerSigningKey = true,
        IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(token.Secret)),
        ValidIssuer = token.Issuer,
        ValidAudience = token.Audience,
        NameClaimType = JwtClaimTypes.Name,
        RoleClaimType = JwtClaimTypes.Roles, // <---- Overriding, new value "role"
    };
});
```

TokenService.cs
```
public class TokenService : ITokenService
{
    private TokenConfigs _tokenManagement;

    public TokenService(IOptions<TokenConfigs> tokenManagement)
    {
        _tokenManagement = tokenManagement.Value;
    }

    public string GetAuthToken(TokenRequest req)
    {
        var claims = new[]
        {
            new Claim(JwtClaimTypes.Name, req.UserName),
            new Claim(JwtClaimTypes.Role, "xxxx"), // <----- Custom type
        };
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_tokenManagement.Secret));
        var credentials = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var jwtToken = new JwtSecurityToken(
            claims: claims,
            issuer: _tokenManagement.Issuer,
            audience: _tokenManagement.Audience,
            expires: DateTime.Now.Date.AddDays(1),
            signingCredentials: credentials
        );
        var token = new JwtSecurityTokenHandler().WriteToken(jwtToken);

        return token;
    }
}
```

AuthController.cs
```
[AllowAnonymous]
[HttpPost]
public async Task<ActionResult<string>> Post([FromBody] TokenRequest req)
{
    string token = null;
    if (isValidUser)
    {
        token = _tokenService.GetAuthToken(req);
    }

    return Ok(token); // <--- Correct result returned
}

[HttpGet]
[Authorize(Roles = "admin,xxxx")] <------ Fail
public ActionResult<string> Get()
{
    var name = User.Identity.Name;
    var isAdmin = User.IsInRole("admin");
    var isXXXX = User.IsInRole("xxxx"); <---- False
    return Ok($"{name} {isAdmin} {isXXXX}");
}
```

3. With these arguments '....'
N/A

Fail to authorize. User.Claims contains incorrect claim type.
```
{http://schemas.microsoft.com/ws/2008/06/identity/claims/role: xxxx}
```

### Expected behavior
Able to pass authorization. Claims should respect what's passed in.

### Screenshots
N/A

### Additional context
Include the output of `dotnet --info`

```
.NET Core SDK (reflecting any global.json):
 Version:   2.2.204
 Commit:    8757db13ec

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.204\

Host (useful for support):
  Version: 2.2.5
  Commit:  0a3c9209c0

.NET Core SDKs installed:
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.400 [C:\Program Files\dotnet\sdk]
  2.1.401 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.502 [C:\Program Files\dotnet\sdk]
  2.1.503 [C:\Program Files\dotnet\sdk]
  2.1.504 [C:\Program Files\dotnet\sdk]
  2.1.505 [C:\Program Files\dotnet\sdk]
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.1.604 [C:\Program Files\dotnet\sdk]
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.2.204 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3-servicing-26724-03 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

**Comment by GeoK** [comment]
Hi @clauzf - the issue here is the JwtSecurityTokenHandler performs inbound 'type' mapping.

You can disable the mapping on the application level:
`JwtSecurityTokenHandler.DefaultMapInboundClaims = false;`

or per instance (assuming that you are using default _SecurityTokenValidator_(s)):

```
...

.AddJwtBearer(x =>
{
    if (x.SecurityTokenValidators.FirstOrDefault() is JwtSecurityTokenHandler jwtSecurityTokenHandler)
        jwtSecurityTokenHandler.MapInboundClaims = false;
});
```

**Comment by brockallen** [comment]
does anyone ever replace the security token validator in either the asp.net core oidc or jwt authentication handlers? in short, this should (have always been) a first class citizen (or never done in the first place).

**Comment by GeoK** [comment]
Hi @brockallen - users can generally replace the token handler and hence I created the code sample in this way.
For example, one of the ideas is to include Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler (which doesn't have any claim mappings) to ASP.NET Core handlers and allow users to choose between System.IdentityModel.Tokens.JwtSecurityTokenHandler and JsonWebTokenHandler.

**Comment by clauzf** [comment]
> 
> 
> Hi @clauzf - the issue here is the JwtSecurityTokenHandler performs inbound 'type' mapping.
> 
> You can disable the mapping on the application level:
> `JwtSecurityTokenHandler.DefaultMapInboundClaims = false;`
> 
> or per instance (assuming that you are using default _SecurityTokenValidator_(s)):
> 
> ```
> ...
> 
> .AddJwtBearer(x =>
> {
>     if (x.SecurityTokenValidators.FirstOrDefault() is JwtSecurityTokenHandler jwtSecurityTokenHandler)
>         jwtSecurityTokenHandler.MapInboundClaims = false;
> });
> ```

Thanks GeoK, it works. Though I still found there is a bit inconsistency, since without your codes, the "name" claim type won't be mapped into internal claim type, only "role" claim type gets mapped. 

Also I'm curious about the syntax you used, `x.SecurityTokenValidators.FirstOrDefault() is JwtSecurityTokenHandler` did a boolean check, then you can still assign the `FirstOrDefault` result to `jwtSecurityTokenHandler`? It works, but feels like magic, is this some new C# syntax sugar?

**Comment by GeoK** [comment]
@clauzf  - Can you provide more information about the inconsistency? Turning the mapping off will give you back claims unmapped to internal types. 
Setting `NameClaimType = JwtClaimTypes.Name` allows you to get the value of _JwtClaimTypes.Name_ claim with `ClaimsIdentity.Name` (_User.Identity.Name_).

To answer your other question, [the is type pattern expression](https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching#the-is-type-pattern-expression) is added in C# 7.0.

**Comment by clauzf** [comment]
Hi @GeoK that's exactly my point.

With my original codes. I didn't do `DefaultMapInboundClaims = false` and overrided both "name" and "role" claim type names. When I read the inbound token only gets "role" claim type mapped to built-in `ClaimTypes.Role` and role related authorization functions not working. But not the "name" and the value of "name" claim type mapped to `ControllerBase.User.Identity.Name` correctly.

If the behaviour is consistent. Both "name" and "role" should map to internal claim type names without setting `MapInboundClaims` to false. And related functions should be broken.

In other words. With `MapInboundClaims == True`, and token `{ "name": "xxxx", "role": "yyyy" }`
While reading inbound token...

Actual: `{ "name": "xxxx", "http://schemas.microsoft.com/ws/2008/06/identity/claims/role": "yyyy" }`
`User.Identity.Name == "xxxx"`, `User.Identity.IsInRole("yyyy") == False`

Expect: `{ "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name": "xxxx",  "http://schemas.microsoft.com/ws/2008/06/identity/claims/role": "yyyy" }`
`User.Identity.Name == null`, `User.Identity.IsInRole("yyyy") == False`


PS: Thanks for the syntax explanation.

**Comment by GeoK** [comment]
@clauzf - Thank you for the summary. 
This is by design as there is no "name" [claim type mapping](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/rel/5.5.0/src/System.IdentityModel.Tokens.Jwt/ClaimTypeMapping.cs).

**Comment by clauzf** [comment]
@GeoK Got it. Still it seems a bit weird to me to have this inbound token claim mapping. What's the purpose of this design? Normally I would expect what the framework read is what has been sent.

**Comment by GeoK** [comment]
@clauzf  - This is legacy code, and making changes to the default mapping would likely break a lot of existing apps that depend on the JwtSecurityTokenHandler.
We are working on including JsonWebTokenHandler (which doesn't have any claim mappings) to asp.net.

**Comment by clauzf** [comment]
@GeoK Thx for the explanation. Feel free to close the ticket. I have no more questions.
